File: examples/skills-as-resources/typescript/src/skill-discovery.ts
/**
 * Skill discovery, content loading, and document scanning module.
 *
 * Discovers Agent Skills by scanning a directory for subdirectories
 * containing SKILL.md files, parses YAML frontmatter for metadata,
 * scans for supplementary documents, and provides secure content loading.
 *
 * Inspired by:
 * - skilljack-mcp by Ola Hungerford (https://github.com/olaservo/skilljack-mcp)
 * - skills-over-mcp by Keith Groves (https://github.com/keithagroves/skills-over-mcp)
 */

import * as fs from "node:fs";
import * as path from "node:path";
import { parse as parseYaml } from "yaml";
import type { SkillMetadata, SkillDocument } from "./types.js";
import { getMimeType } from "./resource-helpers.js";

/** Maximum file size for skill files (1MB). */
const MAX_FILE_SIZE = 1 * 1024 * 1024;

/**
 * Parse YAML frontmatter from SKILL.md content.
 * Expects content to start with --- and have a closing ---.
 */
function parseFrontmatter(content: string): {
  frontmatter: Record<string, unknown>;
  body: string;
} {
  if (!content.startsWith("---")) {
    throw new Error("SKILL.md must start with YAML frontmatter (---)");
  }

  const parts = content.split("---");
  if (parts.length < 3) {
    throw new Error("SKILL.md frontmatter not properly closed with ---");
  }

  const frontmatter = parseYaml(parts[1]) as Record<string, unknown>;
  if (typeof frontmatter !== "object" || frontmatter === null) {
    throw new Error("SKILL.md frontmatter must be a YAML mapping");
  }

  const body = parts.slice(2).join("---").trim();
  return { frontmatter, body };
}

/**
 * Check if a resolved path is within the allowed base directory.
 * Uses fs.realpathSync to resolve symlinks and prevent escape attacks.
 */
export function isPathWithinBase(
  targetPath: string,
  baseDir: string
): boolean {
  try {
    const realBase = fs.realpathSync(baseDir);
    const realTarget = fs.realpathSync(targetPath);
    const normalizedBase = realBase + path.sep;
    return realTarget === realBase || realTarget.startsWith(normalizedBase);
  } catch {
    // Fall back to resolve check if realpathSync fails
    const normalizedBase = path.resolve(baseDir) + path.sep;
    const normalizedPath = path.resolve(targetPath);
    return normalizedPath.startsWith(normalizedBase);
  }
}

/**
 * Recursively scan a directory for files, returning SkillDocument entries.
 * Security: applies path traversal checks and file size limits.
 */
function scanDir(
  dirPath: string,
  relativeTo: string,
  baseDir: string
): SkillDocument[] {
  const documents: SkillDocument[] = [];

  if (!fs.existsSync(dirPath)) return documents;

  let entries: fs.Dirent[];
  try {
    entries = fs.readdirSync(dirPath, { withFileTypes: true });
  } catch {
    return documents;
  }

  for (const entry of entries) {
    const fullPath = path.join(dirPath, entry.name);

    // Security: verify path stays within the skills directory
    if (!isPathWithinBase(fullPath, baseDir)) continue;

    if (entry.isFile()) {
      try {
        const stat = fs.statSync(fullPath);
        if (stat.size > MAX_FILE_SIZE) continue;

        const relativePath = path.relative(relativeTo, fullPath).replace(/\\/g, "/");
        documents.push({
          path: relativePath,
          mimeType: getMimeType(entry.name),
          size: stat.size,
        });
      } catch {
        // Skip files we can't stat
      }
    } else if (entry.isDirectory()) {
      // Recurse into subdirectories
      documents.push(...scanDir(fullPath, relativeTo, baseDir));
    }
  }

  return documents;
}

/**
 * Scan a skill directory for supplementary documents.
 * Finds all files in subdirectories of the skill directory,
 * excluding SKILL.md itself.
 */
export function scanDocuments(
  skillDir: string,
  baseDir: string
): SkillDocument[] {
  const documents: SkillDocument[] = [];

  let entries: fs.Dirent[];
  try {
    entries = fs.readdirSync(skillDir, { withFileTypes: true });
  } catch {
    return documents;
  }

  for (const entry of entries) {
    if (!entry.isDirectory()) continue;
    const subDirPath = path.join(skillDir, entry.name);
    documents.push(...scanDir(subDirPath, skillDir, baseDir));
  }

  return documents;
}

/**
 * Discover all skills in a directory.
 * Scans for immediate subdirectories containing SKILL.md (or skill.md) files,
 * and scans for supplementary documents in each skill directory.
 *
 * Security: Skips files larger than MAX_FILE_SIZE, validates frontmatter fields.
 */
export function discoverSkills(skillsDir: string): Map<string, SkillMetadata> {
  const skillMap = new Map<string, SkillMetadata>();
  const resolvedDir = path.resolve(skillsDir);

  if (!fs.existsSync(resolvedDir)) {
    console.error(`Skills directory not found: ${resolvedDir}`);
    return skillMap;
  }

  const entries = fs.readdirSync(resolvedDir, { withFileTypes: true });

  for (const entry of entries) {
    if (!entry.isDirectory()) continue;

    const skillDir = path.join(resolvedDir, entry.name);

    // Find SKILL.md (prefer uppercase, accept lowercase)
    let skillMdPath: string | null = null;
    for (const name of ["SKILL.md", "skill.md"]) {
      const candidate = path.join(skillDir, name);
      if (fs.existsSync(candidate)) {
        skillMdPath = candidate;
        break;
      }
    }

    if (!skillMdPath) continue;

    // Security: check file size before reading
    const stat = fs.statSync(skillMdPath);
    if (stat.size > MAX_FILE_SIZE) {
      console.error(
        `Skipping ${skillMdPath}: file size ${(stat.size / 1024 / 1024).toFixed(2)}MB exceeds limit`
      );
      continue;
    }

    // Security: verify path is within skills directory
    if (!isPathWithinBase(skillMdPath, resolvedDir)) {
      console.error(`Skipping ${skillMdPath}: path escapes skills directory`);
      continue;
    }

    try {
      const content = fs.readFileSync(skillMdPath, "utf-8");
      const { frontmatter } = parseFrontmatter(content);

      const name = frontmatter.name;
      const description = frontmatter.description;

      if (typeof name !== "string" || !name.trim()) {
        console.error(`Skill at ${skillDir}: missing or invalid 'name' field`);
        continue;
      }
      if (typeof description !== "string" || !description.trim()) {
        console.error(
          `Skill at ${skillDir}: missing or invalid 'description' field`
        );
        continue;
      }

      // Extract optional metadata fields
      const metadata: Record<string, string> = {};
      if (
        frontmatter.metadata &&
        typeof frontmatter.metadata === "object"
      ) {
        for (const [k, v] of Object.entries(
          frontmatter.metadata as Record<string, unknown>
        )) {
          if (typeof v === "string") {
            metadata[k] = v;
          }
        }
      }

      const trimmedName = name.trim();
      if (skillMap.has(trimmedName)) {
        console.error(
          `Warning: Duplicate skill name "${trimmedName}" at ${skillMdPath} — keeping first`
        );
        continue;
      }

      // Scan for supplementary documents
      const documents = scanDocuments(skillDir, resolvedDir);

      skillMap.set(trimmedName, {
        name: trimmedName,
        description: description.trim(),
        path: skillMdPath,
        skillDir,
        metadata: Object.keys(metadata).length > 0 ? metadata : undefined,
        documents,
      });
    } catch (error) {
      console.error(`Failed to parse skill at ${skillDir}:`, error);
    }
  }

  return skillMap;
}

/**
 * Load the full content of a SKILL.md file.
 *
 * Security: Validates that the path is within the skills directory,
 * only reads .md files, and enforces a file size limit.
 */
export function loadSkillContent(
  skillPath: string,
  skillsDir: string
): string {
  // Security: only allow .md files
  if (!skillPath.endsWith(".md")) {
    throw new Error("Only .md files can be read");
  }

  // Security: verify path is within skills directory
  if (!isPathWithinBase(skillPath, skillsDir)) {
    throw new Error("Path escapes the skills directory");
  }

  // Security: check file size
  const stat = fs.statSync(skillPath);
  if (stat.size > MAX_FILE_SIZE) {
    throw new Error(
      `File size ${(stat.size / 1024 / 1024).toFixed(2)}MB exceeds ${(MAX_FILE_SIZE / 1024 / 1024).toFixed(0)}MB limit`
    );
  }

  return fs.readFileSync(skillPath, "utf-8");
}

/**
 * Load a supplementary document from a skill directory.
 *
 * Security: Validates that the path is within the skills directory,
 * rejects path traversal attempts, and enforces a file size limit.
 */
export function loadDocument(
  skill: SkillMetadata,
  documentPath: string,
  skillsDir: string
): string {
  // Security: reject path traversal attempts
  if (documentPath.includes("..")) {
    throw new Error("Path traversal not allowed");
  }

  const fullPath = path.join(skill.skillDir, documentPath);

  // Security: verify path is within skills directory
  if (!isPathWithinBase(fullPath, skillsDir)) {
    throw new Error("Path escapes the skills directory");
  }

  // Security: check file size
  const stat = fs.statSync(fullPath);
  if (stat.size > MAX_FILE_SIZE) {
    throw new Error(
      `File size ${(stat.size / 1024 / 1024).toFixed(2)}MB exceeds ${(MAX_FILE_SIZE / 1024 / 1024).toFixed(0)}MB limit`
    );
  }

  return fs.readFileSync(fullPath, "utf-8");
}

File: examples/skills-as-tools/python/src/skills_as_tools/skill_discovery.py
"""
Skill discovery and content loading module.

Discovers Agent Skills by scanning a directory for subdirectories
containing SKILL.md files, parses YAML frontmatter for metadata,
and provides secure content loading.

Inspired by:
- skilljack-mcp by Ola Hungerford (https://github.com/olaservo/skilljack-mcp)
- skills-over-mcp by Keith Groves (https://github.com/keithagroves/skills-over-mcp)
"""

from __future__ import annotations

import logging
from dataclasses import dataclass, field
from pathlib import Path

import yaml

logger = logging.getLogger(__name__)

# Maximum file size for skill files (1MB)
MAX_FILE_SIZE = 1 * 1024 * 1024


@dataclass
class SkillMetadata:
    """Metadata extracted from a skill's SKILL.md YAML frontmatter."""

    name: str
    description: str
    path: str  # Absolute path to the SKILL.md file
    metadata: dict[str, str] = field(default_factory=dict)


def _parse_frontmatter(content: str) -> tuple[dict, str]:
    """Parse YAML frontmatter from SKILL.md content.

    Returns (frontmatter_dict, body_text).
    """
    if not content.startswith("---"):
        raise ValueError("SKILL.md must start with YAML frontmatter (---)")

    parts = content.split("---")
    if len(parts) < 3:
        raise ValueError("SKILL.md frontmatter not properly closed with ---")

    # Use safe_load to prevent arbitrary code execution
    frontmatter = yaml.safe_load(parts[1])
    if not isinstance(frontmatter, dict):
        raise ValueError("SKILL.md frontmatter must be a YAML mapping")

    body = "---".join(parts[2:]).strip()
    return frontmatter, body


def _is_path_within_base(target: Path, base: Path) -> bool:
    """Check if a resolved path is within the allowed base directory."""
    try:
        resolved_base = base.resolve(strict=True)
        resolved_target = target.resolve(strict=True)
        return resolved_target == resolved_base or str(resolved_target).startswith(
            str(resolved_base) + "/"
        ) or str(resolved_target).startswith(str(resolved_base) + "\\")
    except OSError:
        # Fall back to non-strict resolve
        resolved_base = base.resolve()
        resolved_target = target.resolve()
        return str(resolved_target).startswith(str(resolved_base) + "/") or str(
            resolved_target
        ).startswith(str(resolved_base) + "\\")


def discover_skills(skills_dir: str) -> dict[str, SkillMetadata]:
    """Discover all skills in a directory.

    Scans for immediate subdirectories containing SKILL.md files.
    Security: skips files larger than MAX_FILE_SIZE, validates frontmatter.
    """
    skill_map: dict[str, SkillMetadata] = {}
    resolved_dir = Path(skills_dir).resolve()

    if not resolved_dir.is_dir():
        logger.error("Skills directory not found: %s", resolved_dir)
        return skill_map

    for entry in resolved_dir.iterdir():
        if not entry.is_dir():
            continue

        # Find SKILL.md (prefer uppercase, accept lowercase)
        skill_md_path = None
        for name in ("SKILL.md", "skill.md"):
            candidate = entry / name
            if candidate.exists():
                skill_md_path = candidate
                break

        if skill_md_path is None:
            continue

        # Security: check file size before reading
        stat = skill_md_path.stat()
        if stat.st_size > MAX_FILE_SIZE:
            logger.error(
                "Skipping %s: file size %.2fMB exceeds limit",
                skill_md_path,
                stat.st_size / 1024 / 1024,
            )
            continue

        # Security: verify path is within skills directory
        if not _is_path_within_base(skill_md_path, resolved_dir):
            logger.error(
                "Skipping %s: path escapes skills directory", skill_md_path
            )
            continue

        try:
            content = skill_md_path.read_text(encoding="utf-8")
            frontmatter, _body = _parse_frontmatter(content)

            name = frontmatter.get("name")
            description = frontmatter.get("description")

            if not isinstance(name, str) or not name.strip():
                logger.error(
                    "Skill at %s: missing or invalid 'name' field", entry
                )
                continue
            if not isinstance(description, str) or not description.strip():
                logger.error(
                    "Skill at %s: missing or invalid 'description' field", entry
                )
                continue

            # Extract optional metadata
            extra_metadata: dict[str, str] = {}
            raw_meta = frontmatter.get("metadata")
            if isinstance(raw_meta, dict):
                for k, v in raw_meta.items():
                    if isinstance(v, str):
                        extra_metadata[k] = v

            skill_name = name.strip()
            if skill_name in skill_map:
                logger.warning(
                    "Duplicate skill name '%s' at %s — keeping first",
                    skill_name,
                    skill_md_path,
                )
                continue

            skill_map[skill_name] = SkillMetadata(
                name=skill_name,
                description=description.strip(),
                path=str(skill_md_path),
                metadata=extra_metadata if extra_metadata else {},
            )
        except (OSError, ValueError) as exc:
            logger.error("Failed to parse skill at %s: %s", entry, exc)

    return skill_map


def load_skill_content(skill_path: str, skills_dir: str) -> str:
    """Load the full content of a SKILL.md file.

    Security: validates path is within skills directory, only reads .md files,
    and enforces a file size limit.
    """
    target = Path(skill_path)
    base = Path(skills_dir)

    # Security: only allow .md files
    if target.suffix.lower() != ".md":
        raise ValueError("Only .md files can be read")

    # Security: verify path is within skills directory
    if not _is_path_within_base(target, base):
        raise ValueError("Path escapes the skills directory")

    # Security: check file size
    stat = target.stat()
    if stat.st_size > MAX_FILE_SIZE:
        raise ValueError(
            f"File size {stat.st_size / 1024 / 1024:.2f}MB exceeds "
            f"{MAX_FILE_SIZE / 1024 / 1024:.0f}MB limit"
        )

    return target.read_text(encoding="utf-8")
